### DISCUSSION TITLE
Export "stuff" from egui into png

### DISCUSSION THREAD

**Discussion by fleexo** [emilk/egui#2117] [Q&A] [answered]
Hey,
I want to achieve some kind of export of elements inside of a given container to PNG, but don't have any idea where to start.

I really don't have defined the term "elements" yet so it could be everything. Like custom widgets or shapes (drawn with for example the egui painter).
I thought about exporting all elements (most likely drawn with painter()) in a window (I think this is a suitable container in egui?) to PNG.

Is there any way I could achieve such exportation?

It isn't a full though-out idea yet, so I only could describe the general idea, maybe someone can lead me in the right direction. While reading the documentation of egui I couldn't find anything which fits my needs :D

**Comment by felipetesc** [comment, marked as answer]
I haven't done yet, but eventually I'll face the same problem. Before finding your question I considered to work more or less like this: 

If I want to export as an image:
- I'll create a matrix of pixels (u32), every pixel should have a color 
- I'll convert to a buffer
- use the rust crate image 
- figure out the process of conversion from the buffer to the needed image format


If I want to save as vector:

- create a struct called vector
- each vector struct should have properties to fully reproduce the shape in question to be serialized and deserialized in the future
- every shape should have it's own layer
- I'll have a vector of shapes
- save all shapes inside a file (can be a svg)

**Reply by fleexo** [reply]
Hey, thanks for sharing your idea!

But how do you create the pixel matrix out of the egui element (e.g. painter)?
Or is your approach to first create some figure, independent of egui, which is stored in this matrix and then draw this matrix (as image) to the egui container?

**Reply by felipetesc** [reply]
Please consider this:  this example uses a canvas component because it allows to easily edit the image

Now I can see how poor it was my previous post. Sorry.
(However I only wrote about exporting and not about the rendering process to preview a certain image, which wasn't  requested)

There is an example at https://egui.rs about how to draw using the canvas component. Inside the source code, at https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/painting.rs, emilk uses, to save the pixels, the code at the line 7 ` lines: Vec<Vec<Pos2>>,`  which is a matrix. The point there is to transform the image into this variable called lines. 

To be succint about emilks's demo, his code capture the on press event (1), get the cursor position (2), fill the lines variable (3), and draw at the correct position (Pos2) the right color on the screen.

Differently from his code, what we need to do to is to open a given image, convert its contents into the variable called lines. This can be achieved using an external crate called image available at https://crates.io/crates/image. After filling the lines variable, we need to draw the image itself inside the component called canvas. To do this we can use his code, if we change a litle bit. The lines which are interesting to us can be found between the line 59 and 65. The point here is to map our matriz to canvas:

```rust
for line in &self.lines {
      if line.len() >= 2 {
          let points: Vec<Pos2> = line.iter().map(|p| to_screen * *p).collect();
          shapes.push(egui::Shape::line(points, self.stroke));
      }
  }
```
Once we saved the contents inside the lines variable, the process described previously code can be used with some changes. 

Some helpful posts:
Render buffer as image - https://github.com/emilk/egui/discussions/679
Creating images - https://github.com/emilk/egui/discussions/343
Add Icon Support - https://github.com/emilk/egui/pull/193#issuecomment-900669397
How to display an image in real-time using Rust - https://stackoverflow.com/questions/70317456/how-to-display-an-image-in-real-time-using-rust
Serialize and deserialize u8 and i32 - https://github.com/second-state/serialize_deserialize_u8_i32
Converting an image into a 2D matrix - https://users.rust-lang.org/t/converting-an-image-into-a-2d-matrix/56032
I created a simple image editor using OpenGL and egui - https://www.reddit.com/r/rust/comments/wzs905/media_i_created_a_simple_image_editor_using/
Simp - https://github.com/Kl4rry/simp
